src/rust/pipeline-ingress/src/main.rs

```diff
@@ -0,0 +1,209 @@
+    pipeline::{
+        v1beta1::{
+            Metadata,
+            RawLog,
+        },
+        v1beta2::Envelope,
+    },
+};
+use thiserror::Error;
+use tokio::net::TcpListener;
+use tracing_subscriber::{
+    prelude::*,
+    EnvFilter,
+};
+use uuid::Uuid;
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+enum IngressApiError {
+    #[error("failed to send message to kafka {0}")]
+    ProducerError(#[from] ProducerError),
+}
+
+struct IngressApi {
+    producer: Producer<Envelope<RawLog>>,
+}
+
+impl IngressApi {
+    fn new(producer: Producer<Envelope<RawLog>>) -> Self {
+        IngressApi { producer }
+    }
+}
+
+#[async_trait::async_trait]
+impl PipelineIngressApi<IngressApiError> for IngressApi {
+    #[tracing::instrument(skip(self))]
+    async fn publish_raw_log(
+        &self,
+        request: PublishRawLogRequest,
+    ) -> Result<PublishRawLogResponse, IngressApiError> {
+        let created_time = SystemTime::now();
+        let last_updated_time = created_time;
+        let tenant_id = request.tenant_id;
+        let event_source_id = request.event_source_id;
+        // TODO: trace_id should be generated at the edge. This service is
+        // currently "the edge" but that won't be true forever. When there is an
+        // actual edge service, that service should be responsible for
+        // generating the trace_id.
+        let trace_id = Uuid::new_v4();
+        self.producer
+            .send(Envelope::new(
+                Metadata::new(
+                    tenant_id,
+                    trace_id,
+                    0,
+                    created_time,
+                    last_updated_time,
+                    event_source_id,
+                ),
+                RawLog::new(request.log_event),
+            ))
+            .await?;
+
+        Ok(PublishRawLogResponse::ok())
+    }
+}
+
+#[non_exhaustive]
+#[derive(Debug, Error)]
+enum ConfigurationError {
+    #[error("failed to configure kafka client {0}")]
+    Kafka(#[from] KafkaConfigurationError),
+
+    #[error("failed to configure gRPC server {0}")]
+    Server(#[from] ServerConfigurationError),
+
+    #[error("missing environment variable {0}")]
+    EnvironmentVariable(#[from] VarError),
+
+    #[error("failed to bind socket address {0}")]
+    SocketError(#[from] std::io::Error),
+
+    #[error("failed to parse integer value {0}")]
+    ParseInt(#[from] ParseIntError),
+
+    #[error("failed to configure tracing {0}")]
+    Tracing(#[from] TraceError),
+}
+
+#[tracing::instrument(err)]
+async fn handler() -> Result<(), ConfigurationError> {
+    let socket_address = std::env::var("PIPELINE_INGRESS_BIND_ADDRESS")?;
+
+    let healthcheck_polling_interval_ms =
+        std::env::var("PIPELINE_INGRESS_HEALTHCHECK_POLLING_INTERVAL_MS")?.parse()?;
+
+    let bootstrap_servers = std::env::var("KAFKA_BOOTSTRAP_SERVERS")?;
+    let sasl_username = std::env::var("KAFKA_SASL_USERNAME")?;
+    let sasl_password = std::env::var("KAFKA_SASL_PASSWORD")?;
```

Comment:
These seem like they might be better off in a structopt configuration struct?

Reply:
I think we can (and probably _should_) punt on this type of thing until we figure out a good service abstraction. In particular, it appears `structopt` is deprecated. Their documentation suggests that we use `clap` instead. Additionally, we'll need to do a bunch of work to consolidate (and tease apart, where needed) the protobuf and kafka abstractions to the point that they can work with some kind of a "grapl service builder" type thing. Moreover, we need an idea of service lifecycle a la https://github.com/google/guava/wiki/ServiceExplained, signal handling, logging/tracing configuration, etc.

There are some clear answers here (like we should definitely use tower because it interoperates with tonic nicely) but the finer details of how all this stuff should coalesce aren't really there yet.

I'll work on an RFC for this type of service abstraction moving forward, but I think for now it's best to resist the temptation to formalize it in code just yet.